RFC-0025-improving-incremental-builds.md

```diff
@@ -0,0 +1,217 @@
+This should be extremely rare though.
+
+The initial migration can be assisted by
+[`include-what-you-use`](https://include-what-you-use.org/) which is a
+clang-based tool to find the necessary headers based on what symbols
+are actually used in a file. For most cpp files, it is as simple as
+copying the `<ATen/ops/*>` headers from the tool output.
+
+`TORCH_ASSERT_NO_OPERATORS` cannot be enforced as strictly because
+compliance often requires significantly restructuring the code.
+However, where the cost-benefit is favorable, this macro should be
+defined manually in the source file. This will include files that
+already use `TensorIterator` exclusively and just need their includes
+pruned; and more costly to compile files where taking time to
+restructure them is worthwhile.
+
+## **Metrics**
+
+Some key metrics that should be improved by these changes:
+- Time to rebuild after adding editing a functional operator
+- Time to rebuild after adding editing a method operator
+- `sccache` miss rate in open source CI
+
+## **Drawbacks**
+There is no denying that it's more convenient to just include
+`<ATen/ATen.h>` at the top of the file and not have to think about it.
+This will require additional effort on the part of developers to
+maintain the list of operator includes.
+
+## **Alternatives**
+[pytorch/pytorch#61915](https://github.com/pytorch/pytorch/pull/61915)
+proposed generating new operators into separate headers called
+`*_Delta.h` (e.g. `Functions_Delta.h`) during local development. This
+improves incremental builds locally with comparatively fewer changes
+to the ATen codebase. However,
+
+1. It only helps when adding operators, not changing existing ones
+2. It does nothing for clean builds using a cache (like in CI)
+3. Incremental builds no long reproduce a from-scratch build
+
+[pytorch/pytorch#62258](https://github.com/pytorch/pytorch/issues/62258)
+proposed removing operator header dependencies from CUDA kernels
+specifically. This is aligned with the aims of `TensorBase`, but
+leaves out several hundred non-kernel source files. Individually these
+files take around 10 seconds each to compile, but together account for
+2-3 hours of (serial) CPU time.
+
+
+## **Prior Art**
+Not aware of any relevant prior art.
+
+## **How we teach this**
+This change doesn't effect end-users.
+
+For PyTorch developers, documentation should be written on how to
+write code in compliance with the two enforcement macros and linked in
+the corresponding error messages.
+
+## **Unresolved questions**
+- Is it possible/desirable to enforce `TORCH_ASSERT_NO_OPERATORS` automatically?
+  e.g. for all `.cu` files, or all files over a certain compile time.
+- Can `include-what-you-use` completely automate operator includes?
+  Tools exist for strictly managing _all_ includes, but that would be
+  a significant change from existing include style.
+- Is it worth adopting `TensorBase` more widely than just kernels?
```

Comment:
I think we could answer this by seeing how much benefit we will get from doing that? Is there an easy way to see how this will change the metrics discussed above?

Reply:
I don't know of an easy way but some numbers to put it into perspective: 1184 files include `Tensor.h` today which is ~70% of the compile time for ATen and torch. ~40% of operators have some method variant so keeping it this way means very roughly speaking we can expect ~60% of operator changes to have fast incremental builds.

However, I suspect that methods change less often and that more development effort will be going into new operators which are mostly functional-only. So in practice it may be much more the 60% that are fast.